crates/runc-shim/src/cgroup_memory.rs

```diff
@@ -126,6 +137,111 @@ pub async fn register_memory_event(
     Ok(receiver)
 }
 
+fn memory_event_fd(path: &Path) -> Result<inotify::Inotify> {
+    let instance = inotify::Inotify::init(inotify::InitFlags::empty())?;
+
+    let fpath = path.join("memory.events");
+    instance.add_watch(&fpath, inotify::AddWatchFlags::IN_MODIFY)?;
+
+    let evpath = path.join("cgroup.events");
+    instance.add_watch(&evpath, inotify::AddWatchFlags::IN_MODIFY)?;
+
+    Ok(instance)
+}
+
+async fn parse_kv_file(cg_dir: &Path, file: &str) -> Result<HashMap<String, u32>> {
+    let path = cg_dir.join(file);
+    let mut map: HashMap<String, u32> = HashMap::new();
+
+    let file_string = read_to_string(path.clone()).await.map_err(io_error!(
+        e,
+        "open {}.",
+        path.to_string_lossy()
+    ))?;
+
+    for line in file_string.lines() {
+        if let Some((key, val)) = line.split_once(" ") {
+            let val = val.parse::<u32>()?;
+            map.insert(key.to_string(), val);
+        }
+    }
+
+    Ok(map)
+}
+
+pub async fn register_memory_event_v2(key: &str, cg_dir: &Path) -> Result<Receiver<String>> {
+    let (sender, receiver) = mpsc::channel(128);
+    let cg_dir = Arc::new(Box::from(cg_dir));
+    let key = key.to_string();
+
+    tokio::spawn(async move {
+        let inotify = memory_event_fd(&cg_dir).unwrap();
```

Comment:
You can move this before spawn to handle error properly.
And then move to thread.

Reply:
Also possibly you could extend the function to return a stream of `inotify_event` events to simplify the logic of this loop?

Reply:
@mxpv or to modify here as:
```rust
fn new_read_inotify_file_function(inotify_file) {
         nread = read(inotify_file)
        // verify events here
        if nread >= size_of::<libc::inotify_event>()
        ....
}
pub async fn register_memory_event_v2() {
    // create inotify file
    inotify_file = ...
    tokio::spawn() {
        loop {
            // mv all verifications of events to a new function
            if new_read_inotify_file_function(inotify_file) {
                            parse_kv_file()
                            .......
            } 
        }
    }
}
```